### ISSUE TITLE
[appimaged] Check signatures

### ISSUE THREAD

**Issue by probonopd** [AppImage/appimaged#35]
See `curl -sSL https://get.rvm.io | bash -s stable --ruby` how it informs about missing keys.

**Comment by TheAssassin** [comment]
Why, though? Checking signatures doesn't make for a security model, and doesn't add any trust.

**Comment by TheAssassin** [comment]
That doesn't make for a security model. Unidentified developer doesn't mean unsigned. Their trust model is based on their own PKI, using a hierarchical model. That's nothing that could be applied as-is to AppImages. As said, checking those signatures doesn't create any trust. Trust is created by making use of the existence of valid signatures (checking them is only data this works on, not what makes up the trust).

AppImagteUpdate demonstrates how to _create_ trust using these signatures. If the old and new AppImage are signed, and those signatures are valid and made from the same public key, the new AppImage's trust can be "derived" from the old one.

**Comment by probonopd** [comment]
> Their trust model is based on their own PKI, using a hierarchical model. That's nothing that could be applied as-is to AppImages. 

Which we would have to replace by a web-of-trust instead of a central authority which decides who is trustworthy and who isn't. But as a first step, checking for signatures might possibly be used for matching AppImages to https domains, to verify that an AppImage comes from krita.org. Food for thought.

**Comment by TheAssassin** [comment]
Why do I have to repeat myself on this? Every second day, you _again_ propose this magic "web of trust" and present it as the solution of all problems. It just won't work. Reasons explained in other issues.

PGP keys don't contain domain names. Also, how would you recognize where an AppImage belongs? That information can't be put in the AppImage.

**Comment by probonopd** [comment]
We could require the domain owner to host the public key at a certain URL on the domain, or something like that, to link the two together.

**Comment by TheAssassin** [comment]
That doesn't create any trust yet. The issue is not the distribution of the key, but to create an association between an app in an AppImage and that location where the key is located.
An attacker can just put a key somewhere, and create a fake version with a valid signature.

**Comment by probonopd** [comment]
An attacker would have to take over the web server of the key owner for that.

**Comment by TheAssassin** [comment]
No, that's not the issue here. Try again.

**Comment by probonopd** [comment]
If I am dumb enough to take download-krita-today.freehost.in for krita.org, then well, my fault ;-)

**Comment by TheAssassin** [comment]
That's not the point either. You're working on the wrong bit. Try again. Or let me show you interactively later.

**Comment by TheAssassin** [comment]
TL;DR: To authenticate an AppImage, I need to make sure the key I have is authentic. That problem needs to be solved. None of the existing suggestions can provide that. Therefore, the processes cannot be considered secure.

**Comment by probonopd** [comment]
The easiest way is to just assume that authentic keys are in the user's keychain. Later, we can think about how we can get them there in a non-frightening way.

**Comment by TheAssassin** [comment]
This is the key problem to solve. It can't be simply "deferred" to some point in the future, with a comment `TODO: come up with better and still secure solution`.

**Comment by TheAssassin** [comment]
It's annoying to see the same (bad) ideas come up every couple of months.

The issue with the keyring is less security and more UX related. Answer the following questions (while maintaining a easy and convenient UX):

- How will you get the keys into the keychain?
- How will you validate the keys' authenticity? (No, your proposed system doesn't ensure that, it only allows for validating signatures after the download...)
- How will you remove keys that became insecure? (That's the most important bit, IMO).
- How will you make sure that AppImages are signed with the key that belongs to the project only? (If a key gets stolen, it could be used to sign any AppImage, and all users would accept them.)
- What to do with unsigned AppImages (or old AppImage which don't embed the pubkey yet)?

That's the first five issues that came into my mind that will prevent any signature validation like that in appimaged. And there's plenty more such issues. 

What annoys me is that I often read or hear "we must automate all the things (security related)", and then, when you want to *seriously* discuss such proposals, all these outdated and proven-to-not-work-very-well-with-average-joe that laid in a dusty drawer in some desk in a basement are praised as "the solution". Alternatively, it's either an "implementation detail" or "deferred to a later point".

I don't say there's ultimate security, but if trade-offs are to be made, they need to be well defined, with all their pros and cons. There is no point in implementing insecure or impossible-to-use features, as they either provide no security (which kind of eliminates the need for them), or they won't be used. Both means implementing them is a pure waste of time.

Also, this is not a software security topic. Software security is about the security of _implementations_, not the security of _algorithms_. Nobody would call TLS insecure if a a flaw in an implementation is found, and vice versa.

I know, IT security is a difficult, time consuming job. But if you don't do it right, you can just leave things as they are. Even worse, if you do it wrong and then advertise your solution as secure, users might feel safe while still being vulnerable.

See e.g., [why Telegram is insecure](https://eprint.iacr.org/2015/1177.pdf). Millions of people seem to either not be aware of this, fall for Telegram's advertising (which, of course, propose it's secure), or just don't care. The latter is something IT security folks can not fix (unless they'd finally invent some measures how machines could set up authentic and private communication -> not possible, machines aren't users, and you want to set up the channels between the users, after all). But all the other people would probably not use Telegram if they were aware of all the issues.
And that's just the software issues. Most people don't even switch on proper end-to-end encryption ("private chat" mode); if they don't, all messages are stored in plain text on the Telegram servers. That's far from secure, especially given the background and news articles on them in the last years.
Not to mention that most clients are made by third parties with mostly unverified implementations of that protocol...

More info: https://www.alexrad.me/discourse/a-264-attack-on-telegram-and-why-a-super-villain-doesnt-need-it-to-read-your-telegram-chats.html


This little example showcases why it's a bad idea to implement "something" that isn't secure. And, before you think "but after they've had it they could change to something better": never change a running system, I guess. They earn loads of money with that app _without_ making it secure... Most insecure things won't (or can't) be changed later on. Also a good reason _not_ to implement "something".

**Comment by probonopd** [comment]
You are right in that it is worse to advertise false claims of security than having no security.

**Comment by TheAssassin** [comment]
Exactly. I don't want to build anything half-finished that doesn't quite work or isn't very secure or is inconvenient/impossible to use. Such systems are likely to never be changed any more after the first implementation.

Let's rather try to build a usable concept that provides some real security advantage whose drawbacks are well known. That will require some requirements engineering at first, and shouldn't be done in a rush.

Such a system is e.g., available in AppImageUpdate. We _know_ when it will increase security, and where its drawbacks are (a drawback is that, e.g., unsigned AppImages will be updated showing a warning only; that's a UX decision we made, which reduces the system's overall security, but fits our current situation better).
I can't guarantee the implementation is perfectly secure, but I tried to fix all obvious issues in there, and the sensitive operations are delegated to GPG, so I consider it relatively secure in terms of software security.
The implementation is completely open source, and can be reviewed by anybody. In fact, I'd invite anyone who's got experience in the field of software security to review the code!
